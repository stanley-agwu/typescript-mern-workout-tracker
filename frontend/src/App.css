.App {
  background-color: #eff3f6;
}
