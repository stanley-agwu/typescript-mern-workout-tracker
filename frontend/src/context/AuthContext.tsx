import React, { FC, createContext, useReducer } from 'react';
import { ACTIONS } from '../constants';

import { IUser, UserActions, IUserContext } from '../types';

export const AuthContext = createContext({} as IUserContext);

const initState: IUser = { user: null };
const userReducer = (state: IUser, action: UserActions) => {
  switch(action.type) {
    case ACTIONS.SIGNIN:
      return { user: action.payload };
    case ACTIONS.SIGNOUT:
      return { user: null };
    default:
      return state;
  }
}

export const AuthContextProvider: FC<React.PropsWithChildren> = ({ children }) => {
  //@ts-ignore
  const [state, dispatch] = useReducer(userReducer, initState);

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  )
};