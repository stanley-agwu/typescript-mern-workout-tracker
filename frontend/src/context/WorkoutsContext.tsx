import React, { FC, createContext, useReducer } from "react";
import { ACTIONS } from "../constants";
import { IWorkoutContext, WorkoutActions, IWorkoutState } from "../types";

export const WorkoutsContext = createContext({} as IWorkoutContext);

const initState: IWorkoutState = { workouts: [] };

const workoutReducer = (state: IWorkoutState, action: WorkoutActions) => {
  switch(action.type) {
    case ACTIONS.SET_WORKOUTS:
      return { workouts: action.payload };
    case ACTIONS.CREATE_WORKOUT:
      return { workouts: [action.payload, ...state.workouts]};
    case ACTIONS.UPDATE_WORKOUT:
      //@ts-ignore
      const filteredState = state.workouts.filter((workout) => workout._id !== action.payload._id)
      return { workouts: [action.payload, ...filteredState ] };
    case ACTIONS.DELETE_WORKOUT:
      //@ts-ignore
      return { workouts: state.workouts.filter((workout) => workout._id !== action.payload._id) };
    default: return state;
  }
}

export const WorkoutsContextProvider: FC<React.PropsWithChildren> = ({ children }) => {
  //@ts-ignore
  const [state, dispatch] = useReducer(workoutReducer, initState);
  return (
    <WorkoutsContext.Provider value={{state, dispatch}}>
      {children}
    </WorkoutsContext.Provider>
  )
}