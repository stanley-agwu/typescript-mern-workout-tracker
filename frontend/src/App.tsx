import { BrowserRouter, Routes, Route, Navigate } from 'react-router-dom';
import 'bootstrap/dist/css/bootstrap.min.css';

import './App.css';
import NavBar from './components/Navbar'
import Home from './pages/Home';
import Signin from './pages/Signin';
import Signup from './pages/Signup';
import { useAuthContext } from './hooks/useAuthContext';

const App: React.FC = () => {
  const { state: { user} } = useAuthContext();
  return (
    <div className="App">
      <NavBar />
      <BrowserRouter>
        <Routes>
          <Route path="/" element={user ? <Home /> : <Navigate to="/signin" />} />
          <Route path="/signin" element={!user ? <Signin /> : <Navigate to="/" />} />
          <Route path="/signup" element={!user ? <Signup /> : <Navigate to="/" />} />
          <Route path="*" element={<p>Path not resolved</p>} />
        </Routes>
      </BrowserRouter>
    </div>
  );
}

export default App;
