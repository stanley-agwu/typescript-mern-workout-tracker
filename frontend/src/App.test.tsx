import App from './App';
import { AuthContextProvider } from './context/AuthContext';
import renderer from 'react-test-renderer';

describe('snapshot testing', () => {
  test('snapshot for app component', () => {
    const renderedComponent = renderer.create(
      <AuthContextProvider>
        <App />
      </AuthContextProvider>
    ).toJSON();

    expect(renderedComponent).toMatchSnapshot();
  });
})
