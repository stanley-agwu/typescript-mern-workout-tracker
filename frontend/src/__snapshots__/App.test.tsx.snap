// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing snapshot for app component 1`] = `
<div
  className="App"
>
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark fixed-top"
  >
    <div
      className="container-fluid"
    >
      <a
        className="navbar-brand"
        href="/"
      >
        Freak workouts
      </a>
      <button
        aria-controls="navbarScroll"
        aria-label="Toggle navigation"
        className="navbar-toggler collapsed"
        onClick={[Function]}
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        aria-expanded={null}
        className="navbar-collapse collapse"
        id="navbarScroll"
      >
        <div
          className="me-auto my-2 my-lg-0 navbar-nav navbar-nav-scroll"
          onKeyDown={[Function]}
          style={
            Object {
              "maxHeight": "100px",
            }
          }
        >
          <a
            className="nav-link"
            data-rr-ui-event-key="/"
            disabled={false}
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
          <a
            className="nav-link"
            data-rr-ui-event-key="#"
            href="#"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            About
          </a>
        </div>
        <div
          className="ms-auto my-2 my-lg-0 navbar-nav navbar-nav-scroll"
          onKeyDown={[Function]}
          style={
            Object {
              "maxHeight": "100px",
            }
          }
        >
          <a
            className="nav-link"
            data-rr-ui-event-key="/signin"
            disabled={false}
            href="/signin"
            onClick={[Function]}
          >
            Sign in
          </a>
          <a
            className="mx-4 ps-4 nav-link"
            data-rr-ui-event-key="/signup"
            disabled={false}
            href="/signup"
            onClick={[Function]}
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
